engine/access/ingestion/version/version_control.go

```diff
@@ -0,0 +1,281 @@
+	nodeVersion *semver.Version
+	// versionBeacon stores the last handled version beacon.
+	versionBeacon *flow.SealedVersionBeacon
+
+	// consumers stores the list of consumers for version updates.
+	consumers []VersionControlConsumer
+
+	// blockFinalizedChan is a channel for receiving block finalized events.
+	blockFinalizedChan chan *flow.Header
+
+	// startHeight and endHeight define the height boundaries for version compatibility.
+	startHeight uint64
+	endHeight   uint64
+}
+
+var _ protocol.Consumer = (*VersionControl)(nil)
+var _ component.Component = (*VersionControl)(nil)
+
+// NewVersionControl creates a new VersionControl instance.
+//
+// We currently have no strong guarantee that the node version is a valid semver.
+// See build.SemverV2 for more details. That is why nil is a valid input for node version.
+// Without a node version, the stop control can still be used for manual stopping.
+func NewVersionControl(
+	log zerolog.Logger,
+	versionBeacons storage.VersionBeacons,
+	nodeVersion *semver.Version,
+	latestFinalizedBlock *flow.Header,
+) *VersionControl {
+	// blockFinalizedChan is buffered to handle block finalized events.
+	blockFinalizedChan := make(chan *flow.Header, 1000)
+
+	vc := &VersionControl{
+		log: log.With().
+			Str("component", "version_control").
+			Logger(),
+
+		blockFinalizedChan: blockFinalizedChan,
+		nodeVersion:        nodeVersion,
+		versionBeacons:     versionBeacons,
+		startHeight:        NoHeight,
+		endHeight:          NoHeight,
+	}
+
+	if vc.nodeVersion != nil {
+		log = log.With().
+			Stringer("node_version", vc.nodeVersion).
+			Logger()
+	}
+
+	log.Info().Msg("Created")
+
+	// Setup component manager for handling worker functions.
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(vc.processEvents)
+	cm.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		vc.checkInitialVersionBeacon(ctx, ready, latestFinalizedBlock)
+	})
+
+	vc.Component = cm.Build()
+
+	return vc
+}
+
+// checkInitialVersionBeacon checks the initial version beacon at the latest finalized block.
+// It ensures the component is not ready until the initial version beacon is checked.
+func (v *VersionControl) checkInitialVersionBeacon(
+	ctx irrecoverable.SignalerContext,
+	ready component.ReadyFunc,
+	latestFinalizedBlock *flow.Header,
+) {
+	// component is not ready until we checked the initial version beacon
+	defer ready()
+
+	vb, err := v.versionBeacons.Highest(latestFinalizedBlock.Height)
+	if err != nil {
+		ctx.Throw(
+			fmt.Errorf(
+				"failed to get highest version beacon for version control: %w",
+				err))
+		return
+	}
+
+	if vb == nil {
+		// no version beacon found
+		// this is unexpected as there should always be at least the
+		// starting version beacon, but not fatal.
+		// It can happen if the node starts before bootstrap is finished.
+		// TODO: remove when we can guarantee that there will always be a version beacon
```

Comment:
@janezpodhostnik do we still need this check that the version beacon system is setup? (this was copied from `StopControl`)

Reply:
Yes. This will still be hit before bootstrapping is complete.